seamlessClone bug

Opencv 3.4.6-dev
Python 3.6.8


There is a problem in seamlessClone that was produced by this pr https://github.com/opencv/opencv/pull/12512.

1.  When mask is on borders only (like border line) it's just crashes. It happens because
https://github.com/opencv/opencv/blob/72ccb5fe6e355e2e6e7c6960b5a456c1d08ec28e/modules/photo/src/seamless_cloning.cpp#L77 
this line places zeros on mask borders.
2. It produces undesirable ghost effect on borders

Here is code to reproduce:
```LINES = 60
dst = np.ones((101, 101, 3), np.uint8) * 255
src = np.ones((101, 101, 3), np.uint8) * 255
cv.circle(dst,  center=(50, 50), radius=5, color=(0, 0, 255), thickness=-1)
cv.circle(dst, center=(50, 0), radius=5, color=(0, 0, 255), thickness=-1)

mask = np.zeros((101, 101, 1), np.uint8)
mask[:LINES,:] = 255

res = cv.seamlessClone(src, dst, mask, (50, 50), cv.NORMAL_CLONE)
cv.imwrite('result.png', np.concatenate([dst, src, np.concatenate([mask, mask, mask], axis=2), res], axis=1))
```

That's even stranger behavior 
If i set LINES to 1 or 2 i get
`terminate called after throwing an instance of 'std::length_error'
  what():  vector::_M_default_append
Aborted (core dumped)
`
If i set LINES to 3 i get 
`Traceback (most recent call last):
  File "seamless_test.py", line 13, in <module>
    res = cv.seamlessClone(src, dst, mask, (50, 50), cv.NORMAL_CLONE)
cv2.error: OpenCV(3.4.6-dev) /home/vfilev/packages/opencv/modules/core/src/matrix_wrap.cpp:1658: error: (-215:Assertion failed) !fixedSize() in function 'release'
`

and if i set LINES to 60 i get the result below
Example:  
![result](https://user-images.githubusercontent.com/2921717/63006070-71768a80-be86-11e9-952f-aeaaff4a6278.png)

So you can see that circle inside image is gone but circle on a border is left here.

--- a/modules/photo/src/seamless_cloning.cpp
+++ b/modules/photo/src/seamless_cloning.cpp
@@ -78,6 +78,7 @@ void cv::seamlessClone(InputArray _src, InputArray _dst, InputArray _mask, Point
     copyMakeBorder(mask_inner, mask, 1, 1, 1, 1, BORDER_ISOLATED | BORDER_CONSTANT, Scalar(0));
 
     Rect roi_s = boundingRect(mask);
+    if (roi_s.empty()) return;
     Rect roi_d(p.x - roi_s.width / 2, p.y - roi_s.height / 2, roi_s.width, roi_s.height);
 
     Mat destinationROI = dest(roi_d).clone();

--- a/modules/photo/test/test_cloning.cpp
+++ b/modules/photo/test/test_cloning.cpp
@@ -86,6 +86,15 @@ TEST(Photo_SeamlessClone_normal, regression)
     EXPECT_LE(errorINF, 1);
     double errorL1 = cvtest::norm(reference, result, NORM_L1);
     EXPECT_LE(errorL1, reference.total() * numerical_precision) << "size=" << reference.size();
+
+    mask = Scalar(0, 0, 0);
+    seamlessClone(source, destination, mask, p, result, 1);
+
+    reference = destination;
+    errorINF = cvtest::norm(reference, result, NORM_INF);
+    EXPECT_LE(errorINF, 1);
+    errorL1 = cvtest::norm(reference, result, NORM_L1);
+    EXPECT_LE(errorL1, reference.total() * numerical_precision) << "size=" << reference.size();
 }
 
 TEST(Photo_SeamlessClone_mixed, regression)

